pkg/qemu/qemu.go

```diff
@@ -243,6 +247,62 @@ func showDarwinARM64HVFQEMU620Warning(exe, accel string, features *features) {
 	logrus.Warn(w)
 }
 
+func getMacOSProductVersion() (*semver.Version, error) {
+	cmd := exec.Command("sw_vers", "-productVersion")
+	b, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute %v: %w", cmd.Args, err)
+	}
+	ver, err := semver.NewVersion(string(b))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse macOS version %q: %w", string(b), err)
+	}
+	return ver, nil
+}
+
+// adjustMemBytesDarwinARM64HVF adjusts the memory to be less than 4 GiB, only when the following conditions are met:
+//
+// - memBytes  >= 4 GiB
+// - Host OS   <  macOS 12.4
+// - Host Arch == arm64
+// - Accel     == hvf
+// - QEMU      >= 7.0
+//
+// This adjustment is required for avoiding host kernel panic. The issue was fixed in macOS 12.4 Beta 1.
+// See https://github.com/lima-vm/lima/issues/795 https://gitlab.com/qemu-project/qemu/-/issues/903#note_911000975
+func adjustMemBytesDarwinARM64HVF(memBytes int64, accel string, features *features) int64 {
+	if memBytes < (1 << 32) { // less than 4GiB
+		return memBytes
+	}
+	if runtime.GOOS != "darwin" {
+		return memBytes
+	}
+	if runtime.GOARCH != "arm64" {
+		return memBytes
+	}
+	if accel != "hvf" {
+		return memBytes
+	}
+	if !features.VersionGEQ7 {
+		return memBytes
+	}
+	macOSProductVersion, err := getMacOSProductVersion()
+	if err != nil {
+		logrus.Warn(err)
+		return memBytes
+	}
+	if !macOSProductVersion.LessThan(*semver.New("12.4.0")) {
+		return memBytes
+	}
+	const safeSize = 1 << 31 // 2GiB
```

Comment:
Why is the `safeSize` 2GiB and not e.g. 4GiB - 1 or something quite close to 4GiB?

If you ask for 3GiB, you get 3Gib, but if you ask for 5GiB you only get 2GiB, which seems weird. Shouldn't the default be the same as the highest value that we pass through untouched? If `4GiB - 1` doesn't work, then we should not allow that either.

Reply:
We convert the number into MiB anyways before passing it to qemu, so technically our limit would be 4095MiB.

Reply:
Because I suppose an "odd" memory size might not be "safe" under some conditions.
(I don't have M1 Pro/Max to confirm)

The user can still specify an odd memory size.

Reply:
> Because I suppose an "odd" memory size might not be "safe" under some conditions.

This sounds like a weird reason, as any number of things _might_ not be "safe" either.

I could see the argument that we should change the **default** value to 2GiB if the user doesn't specify anything. But if the user **specifies** 3GiB, they get 3GiB, however, if they **specify** 4GiB, they only get 2GiB, that just feels wrong to me.

Since this is just a temporary workaround under limited conditions, I won't argue this too much, but ideally we should confirm if 4095MiB is safe or not, and use that if it doesn't show the issue.

Reply:
Fixed to 4095 MiB.